storage: prevent removed replica from getting lease

tbg opened the pull request:
Out for early review since the fix seems easier than what has been proposed
in #15385 so far. Will add tests in this PR.

Only review my commits, those of @andreimatei are from #15711 (which is
waiting for @bdarnell as well).

---

Make sure that a lease with an intended lease holder that has since been
removed catches a forcedError.

See #15385.

tbg committed:
storage: simplify lease check in processRaftCommand

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15754)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, though we need @bdarnell's eyes on this as well.

---

Reviewed 1 of 1 files at r1.
Review status: 0 of 31 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15754#-:-Kj_UN-kLSWqt0BCeXP9:b7x2zv3)*
<!-- Sent from Reviewable.io -->

a-robinson commented on pkg/storage/replica.go:
```diff
@@ -980,6 +980,34 @@ func (r *Replica) redirectOnOrAcquireLease(ctx context.Context) (LeaseStatus, *r
+						// submit a proposal for the lease (correctly using as a ProposerLease the
+						// last active lease), and would receive it given the up-to-date
+						// ProposerLease. Hence, an extra check is in order: processRaftCommand
+						// makes sure that leases for a replica not in the descriptor are bounced.
```
I think this would be a little more clear if you replaced s/leases/lease requests/

tschottdorf replied:
Done.

a-robinson reviewed (approved):
LGTM, and thanks for independently re-thinking through all this.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

This change downstream of raft could theoretically cause replica divergence during rolling upgrades, but since it's an assertion against an event that would already cause the range to get stuck (and does not appear to be occurring in practice), it should be OK. 

---

Reviewed 3 of 4 files at r6, 1 of 1 files at r7.
Review status: 4 of 31 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15754#-:-KjcdKZKO1TeGuJ_uIrY:bny263c)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 4 of 31 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15754#-:-Kjdterr_5RXcOg92cvz:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: prevent removed replica from getting lease

Make sure that a lease with an intended lease holder that has since been
removed catches a forcedError.

See #15385.

tbg committed:
storage: add test for removed replica requesting lease & associated refactoring

tschottdorf commented:
Updated with a test. Spent nontrivial time on a higher-level test, but I think this one's as straightforward as we're gonna get.

bdarnell commented:
LGTM

---

Reviewed 3 of 27 files at r8, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15754#-:-KkWdzhOjnNiTd1-SfLZ:bb6ywmx)*
<!-- Sent from Reviewable.io -->
